Add a class to handle updating the payment method after a billing invoice failure

james-allan opened the pull request:
#### 🛑 Blocked status 🛑 

This PR is blocked by the following PRs:

- [x] #2811 
- [x] #2843
- [ ] Requires 26-gh-Automattic/woocommerce-payments-dev-tools for testing.

------
#### Changes proposed in this Pull Request

This PR is porting over the _change payment method class_ developed in our prototype for Stripe Billing. It adds the following functionality.

When a subscription is a WC Pay billing subscription with a failed renewal order we:

1. Override the default 'change payment' method button on the **My Account > Subscriptions > View Subscription** so customers are prompted to update their card details instead.
2. Replace the 'pay' link next to the failed order. This is to prevent customers from trying to pay for the order instead of updating their details. The link now directs the customer to the 'update payment method' page.
3. Redirect customers who manage to land on a pay for order page for an order linked to a WC pay billed subscription redirecting them through the update payment method page flow instead. This catches customers who may have received a link via email for instance. 
4. Changes the change payment method page's title, breadcrumbs and notice content for a more appropriate message. 
 
<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

**Set up:**
1. Install the https://github.com/Automattic/woocommerce-payments-dev-tools plugin
2. Check out the 26-gh-Automattic/woocommerce-payments-dev-tools PR. 
3. Enable the billing clocks functionality following [these steps](https://github.com/Automattic/woocommerce-payments-dev-tools/blob/a21c7093051e3efeb8a5f92af7ba0044b36dcfb0/README.md#billing-clocks).
4. Your site will need to be receiving webhooks from Stripe. https://github.com/Automattic/woocommerce-payments-server/tree/trunk/local#5-listen-to-webhooks

----

1. Purchase a subscription
2. Go to the **WooCommerce > Subscriptions > Edit Subscription** screen on the admin dashboard.
3. Go to the subscription actions dropdown and select **"Set up custom billing clock"**. https://d.pr/i/cFv62X
4. **Save**. 
5. This will set up the billing clock enabled subscription with a new user etc.
6. From the Subscriptions actions dropdown (https://d.pr/i/tLjGQ8) select **`Trigger upcoming invoice`**, **`Trigger invoice created`**,  and **`Fail the next invoice`** in that order, submitting the form each time. _**Note: if the correct action doesn't appear at any stage, refresh the page as there may be some lag time between receiving the response from Stripe.**_
7. In an incognito browser log in as the customer using the instructions found in the updated readme [here](https://github.com/Automattic/woocommerce-payments-dev-tools/blob/a21c7093051e3efeb8a5f92af7ba0044b36dcfb0/README.md#to-pay-for-failed-renewals).
8.  Go to **My Account > My Subscription.** 
    - Here you should find an on-hold subscription with an **'Update payment method'** button option. 
    - If you scroll down there will be a failed renewal with a **'Pay'** link https://d.pr/i/TYNp56
    - Click either the  **'Update payment method'** or **'Pay'** link. They both go to the same page.
    - You should land on a page that looks like a pay for order/change payment method page. https://d.pr/i/02gwnF
    - Enter a new method or choose an existing saved method.  
    - If you entered a successful payment method, the payment should be reattempted and the subscription activated. 
        - The subscription's payment method will also update to the successful card.
        -  You can also verify the subscription's payment method updates by clicking the `sub_xyz` link on the edit subscription screen. https://d.pr/i/OowOIp
        - This [0341](https://d.pr/i/8qqlxd) (failing card number) will be updated to the new card.
    - If you entered a failing payment method, the error will be displayed on the change payment method form and you'll be prompted to try again.

-------------------
 
**Notes**

- This PR also includes an update to the `subscriptions-base` to include the latest changes. This was particularly necessary to test this PR because it includes new actions/filters for changing some Change payment method page titles, and notices.  See: 4162-gh-woocommerce/woocommerce-subscriptions

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
-->

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

aprea commented on tests/unit/subscriptions/test-class-wc-payments-subscription-change-payment-method.php:
```diff
@@ -0,0 +1,192 @@
+		$mock_subscription = new WC_Subscription();
+
+		// Test the false case - no change to input.
+		$this->assertEquals( [], $this->change_payment_method_handler->update_subscription_change_payment_button( [], $mock_subscription ) );
```
Adding a few assertion nitpicks, feel free to ignore them if they're not helpful 😄

`assertSame` is typically a little nicer than `assertEquals`

Reference: paJDYF-17V-p2

james-allan replied:
TIL thanks @aprea!

aprea commented on tests/unit/subscriptions/test-class-wc-payments-subscription-change-payment-method.php:
```diff
@@ -0,0 +1,192 @@
+
+		$result = $this->change_payment_method_handler->update_subscription_change_payment_button( $default_actions, $mock_subscription );
+
+		$this->assertTrue( is_array( $result ) );
```
```suggestion
		$this->assertIsArray( $result );
```

james-allan replied:
It turns out `assertIsArray()` isn't available on at least one of the PHPUnit versions we're using. https://github.com/Automattic/woocommerce-payments/pull/2788/checks?check_run_id=3481575015

How would you usually get around that? Go back to `assertTrue`, or something else like a version check etc?

aprea replied:
Ah, good point! Not much we can do there unless we upgrade PHPUnit across the board, let's stick with `assertTrue` for now 👍

aprea commented on tests/unit/subscriptions/test-class-wc-payments-subscription-change-payment-method.php:
```diff
@@ -0,0 +1,192 @@
+
+		$this->assertTrue( is_array( $result ) );
+		$this->assertContainsOnly( 'array', $result );
+		$this->assertEquals( 2, count( $result ) );
```
```suggestion
		$this->assertCount( 2, $result );
```

james-allan commented:
Updated testing instructions using the Billing Clock integration in the dev tools found here: 26-gh-Automattic/woocommerce-payments-dev-tools

1. Purchase a subscription
2. Go to the **WooCommerce > Subscriptions > Edit Subscription** screen on the admin dashboard.
3. Go to the subscription actions dropdown and select **"Set up custom billing clock"**. https://d.pr/i/cFv62X
4. **Save**. 
5. This will set up the billing clock enabled subscription with a new user etc.
6. From the actions dropdown select **`Trigger upcoming invoice`**, **`Trigger invoice created`**,  and **`Fail the next invoice`** in that order saving each time. Note: if the action doesn't appear at any stage, refresh the page as their may be some lag time between receiving the response.
7. In an incognito browser log in as the customer using the instructions found here:.
8. Go to **My Account > My Subscription.** 
9. Here you should find an on-hold subscription with an Update payment button option. 
10. Click that and change the payment method.  
11. If you've entered a successful payment method, the payment should be reattempted and the subscription activated (may need to reload). 

----
### Further Improvements: 

The following improvements still need to be made to this customer flow.

1. Updating the payment method and the subsequent payment retry relies on the subscription's payment token being updated. Because of that, if the saved method you select is the same as the current one. Nothing happens. The customer lands back on the subscription page with a "payment updated" notice when in fact it hasn't.
2. Because the request to reattempt payment with the new payment method happens after form submit. There is some lag time between when those requests are fired off to Stripe and when Stripe responds with a paid/failed invoice webhook. This means the customer lands on the My Account View Subscription page and the subscription is still on-hold. We may need to apply a temporary lock to the subscription (to prevent the customer from updating the payment again) and display a notice saying something along the lines of we're reattempting the payment hold tight or similar. 

### An important note

While testing I had a lot of difficulty in getting this to work because the token stored on the subscription wasn't updating or sticking. This may be more prominent in testing where we're changing the payment method on the fly, however I got the sense that this isn't currently a robust solution and may fall over if a subscription gets out of sync with the payment method set on Stripe's end. 

I later discovered that the token issue was caused by [this code](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/includes/compat/subscriptions/trait-wc-payment-gateway-wcpay-subscriptions.php#L615). When a subscription is saved it gets the subscriptions token (via `_payment_tokens` meta) and ensures that the `_payment_method_id` meta matches the token. This was overriding some of the billing clock integration. However, it has made me think if we need to remove that code for stripe billing subscriptions and instead keep the payment method ID in sync with what is stored on Stripe's end, not on WC's end.

james-allan commented:
I've just moved this out of the needs review while I fix up the merge conflicts and the description with the latest UX.

james-allan commented:
### An Update

I've made a few improvements since my last comment on this PR. I'll explain how I've addressed the issues I raised in my last comment (https://github.com/Automattic/woocommerce-payments/pull/2788#issuecomment-910026000).

**Removed the requirement on the token changing**

We were previously using the `woocommerce_payment_token_added_to_order` hook to update the payment method at Stripe and then reattempt renewal payment. That hook only triggers though if the token has changed and so if the customer selected the same payment method on the change payment method page, nothing would happen. 

I fixed this issue by introducing a new hook to the WCPay process payment function that we can hook onto when changing a subscriptions payment method even if the token hasn't actually changed. 

The added benefit of this approach is that if the customer has transferred funds, they can reuse the same method and it might now be successful. Changing the token/method/.card isn't actually a requirement for the retry to be successful.

**Handle the retry payment request in the form submit request**

Previously we were sending the request to the server to process a payment request on the failed invoice. We would then wait for the webhook to be sent and received before we completed the order and activated the subscription. This UX meant the customer would submit the change payment form and then land back on the My account subscription page, with a notice about the payment method update, the subscription would still be on-hold and the order would still be failed. We'd receive the webhook, complete the order and activate the subscription but the customer would need to reload the page before seeing all that happen. 

Because I now handle the response in the same form submit request, the subscription is activated right away. The WCPay server will still send the paid webhook, however it will be ignored because we've already handled the payment for that invoice.

This PR should now be ready for review, however, it should be reviewed after #2811 is merged

james-allan commented:
Just making a note that because of the changes I mentioned above, the unit tests are failing because they expect the token update to also send the payment request. Because that no longer happens, I'll need to rework some of the unit tests to reset the expectations.

james-allan commented on includes/subscriptions/class-wc-payments-invoice-service.php:
```diff
@@ -227,7 +227,7 @@ public static function get_pending_invoice_id( WC_Subscription $subscription ) :
 	 * @param WC_Subscription $subscription The subscription.
 	 * @param string          $invoice_id   The invoice ID.
 	 */
-	private function set_pending_invoice_id( WC_Subscription $subscription, string $invoice_id ) {
```
Removing these types from the function signature isn't ideal however they were causing a number of unit testing related issues.

Namely, this [`test_card_is_saved_when_updating_subscription_using_saved_payment_method()`](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/tests/unit/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php#L371) unit test  function creates an order object and passes it to `$this->mock_wcpay_gateway->process_payment( $order->get_id() );` however that order is suppose to behave like a subscription (note the [`WC_Subscriptions::set_wcs_is_subscription()`](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/tests/unit/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php#L381) call) to test changing the payment method. 

That process payment call eventually leads to `maybe_attempt_payment_for_subscription()` which then tries to get the pending invoice by calling `get_pending_invoice_id()` because the object is an order an not a subscription, the type error is thrown.

I tried changing that unit test to instead use the mock `WC_Subscription` instead of an order but that caused different _type_ related issues. Namely, the [`from_payment_request `](https://github.com/Automattic/woocommerce-payments/blob/add/subscription_change_method/includes/class-payment-information.php#L199) expects an `WC_Order` but because in the testing environment the `WC_Subscription` doesn't extend the `WC_Order` that fails. 

I then tried to change the `WC_Subscription` mock class to extend a `WC_Order`, but that caused much broader issues.

If anyone has some ideas how we can keep these argument types, then let me know. 😅

chickenn00dle replied:
I think there is another case of this with [WC_Payments_Invoice_Service::get_subscription_invoice_id](https://github.com/Automattic/woocommerce-payments/blob/7c68a6f407eb8fe0c1497cec6856e85225a39d0c/includes/subscriptions/class-wc-payments-invoice-service.php#L263)

It looks like the [WC_Payment_Gateway_WCPay_Subscriptions_Process_Payment_Test](https://github.com/Automattic/woocommerce-payments/blob/develop/tests/unit/test-class-wc-payment-gateway-wcpay-subscriptions-process-payment.php#L170) tests trigger `maybe_record_first_invoice_payment` which attempts to get the invoice id from the subscription.

I've removed that type hint to get around this as well. I'm wondering if this is the preferred solution, or if we should open an issue for this and see if we can approach how we handle subscriptions in tests differently. What do you think?

haszari replied:
About these type hints (drive by comment!) - if the function supports different types for params, let's make sure the docs are correct, and ideally type hint as well.

mattallan reviewed (approved):
Nice work on this PR @james-allan ! Some good improvements in this PR to the original implementation like actually using the response from retrying the failed invoice to fix up the subscription instead of relying on the webhook 💯 

I'm making this as ready to merge 🍏
